File: Lib/Platform/XOS/Kinematics/test_kinematics.cc
#include "XOSKinematics.h"
#include "Transform.h"
#include <stdio.h>

/*
void print_transform(Transform tr) {
  for (int i = 0; i < 4; i++) {
    for (int j = 0; j < 4; j++) {
      printf("%.4f ",tr(i,j));
    }
    printf("\n");
  }
  printf("\n");
}
*/

int main()
{
  double pLLeg[6], pRLeg[6], pTorso[6];
  std::vector<double> q;

  for (int i = 0; i < 6; i++) {
    pLLeg[i] = 0;
    pRLeg[i] = 0;
    pTorso[i] = 0;
  }

  pLLeg[1] = +.05;
  pRLeg[1] = -.05;
  pTorso[0] = 0.02;
  //  pTorso[2] = 0.30;
  pTorso[2] = 0.20;
  pTorso[3] = -.12;
  pTorso[4] = .13;
  pTorso[5] = .09;

  Transform trTorso = transform6D(pTorso);
  std::vector<double> pTest = position6D(trTorso);
  printf("pTorso: ");
  for (int i = 0; i < 6; i++) printf("%.4f ", pTorso[i]);
  printf("\n");

  //  nao_kinematics_inverse_legs(q, pLLeg, pRLeg, pTorso, 1);
  q = darwinlc_kinematics_inverse_legs(pLLeg, pRLeg, pTorso, 1);

  for (int i = 0; i < 12; i++) {
    printf("%.4f ", q[i]);
  }
  printf("\n");

  /*
  Transform t = nao_kinematics_forward_lleg(q);
  print_transform(inv(t));

  t = nao_kinematics_forward_rleg(q+6);
  print_transform(inv(t));
  */
}

File: Lib/Platform/XOS/Kinematics/test_kinematics.lua
local cwd = '.';
-- Get Computer for Lib suffix
local computer = os.getenv('COMPUTER') or '';
local computer = 'Darwin' -- can get from uname
if (string.find(computer, 'Darwin')) then
  -- MacOS X uses .dylib:
  print('Mac!')
  package.cpath = cwd..'/Lib/?.dylib;' .. package.cpath;
  package.cpath = cwd .. '/?.dylib;' .. package.cpath
else
  print('Comp:', computer)
  package.cpath = cwd .. '/Lib/?.so;' .. package.cpath;
end

kinematics = require('XOSKinematics')

function print_table(t)
  local i;
  local s = "";
  for i = 1,#(t) do 
    s = s..string.format("%.4f ", t[i]);
  end
  print(s);
end

function print_transform(t)
  local i, j;
  local s = "";
  for i = 1,4 do
    for j = 1,4 do
      s = s..string.format("%.4f ",t[i][j]);
    end
    s = s.."\n";
  end
  print(s);
end

pLLeg = {0, .05, 0, 0, 0, 0};
pRLeg = {0, -.05, 0, 0, 0, 0};
--pTorso = {0.015, 0, 0.30, 0, 0, 0};
pTorso = {0.015, 0, 0.21, 0, 0, 0};
pTorso[2] = 0.0;

q = kinematics.inverse_legs(pLLeg, pRLeg, pTorso, 0);
--q[2] = q[2]+.05;
print_table(q);

tLeft = kinematics.forward_lleg({q[1],q[2],q[3],q[4],q[5],q[6]});
print_transform(tLeft);

tRight = kinematics.forward_rleg({q[7],q[8],q[9],q[10],q[11],q[12]});
print_transform(tRight);
